Review and update Dask tutorial (#250)
nebari-dev/nebari-docs: docs/docs/tutorials/using_dask.md

---
id: using_dask
title: Work with big data using Dask
description: Introduction to Dask with Nebari
---

# Work with big data using Dask

## Introduction

Working with large datasets can pose a few challenges - perhaps the most common is memory limitations on the
user's machine.

[Dask](https://www.dask.org/) is a flexible, open source library for parallel and distributed computing in Python.
Dask allows data scientists the ability able to scale analyses from a sample dataset to the full, large-scale dataset
with almost no code changes.
It's a powerful tool that can revolutionize how you do analytics!

## Dask integration on Nebari

Nebari uses [Dask Gateway][dask-gateway] to expose auto-scaling compute clusters automatically

> configured for the user, and it provides a secure way to managing Dask clusters.

Reviewer: We should cross reference the advanced settings doc here https://github.com/nebari-dev/nebari-docs/pull/115/files#diff-7f41f42d3602fd87b565db4e582bb2ff06b2a1da146eab2e602d2594623c3c0e

<details>

> <summary> Click here for quick notes on how this Dask works! </summary>

Reviewer: ```suggestion
<summary> Click here for quick notes on how Dask works! </summary>
```

The `client` interacts with both `scheduler` (sends instructions) and `workers` (collects results).

Check out the [Dask documentation][dask-docs] and the [Dask Gateway documentation][dask-gateway] for a full explanation.

</details>

## Step 1 - Set up Dask Gateway

Let's start with a fresh Jupyter notebook.
Select an environment from the `Select kernel` dropdown menu
(located on the top right of your notebook).

> On a default Nebari deployment, you can select the `filesystem/dask` environment.

Reviewer: Need to double check the default dask env name.

Author: Thanks, updated!

:::warning
Be sure to select an environment which includes `Dask`, and note that the versions of `dask`, `distributed`, and `dask-gateway` must be the same.

> We recommend the [`nebari-dask` metapackage](https://anaconda.org/conda-forge/nebari-dask) that includes the correct Dask packages and versions.

Reviewer: ```suggestion
We recommend the [`nebari-dask` metapackage](https://anaconda.org/conda-forge/nebari-dask). We publish this metapackage alongside Nebari to easily provide you with the correct Dask packages and versions, just be sure that the `nebari-dask` version matches your Nebari deployment version.
```
Note - we should add a note about bumping the nebari-dask version in the "how to bump the nebari version docs"

options
```

> ![Nebari - Cluster Options UI](/img/tutorials/dask_cluster_options.png)

Reviewer: This screenshot needs to be updated pending the question on the name of the default dask env.

Author: Done!

![Dask diagnostic UI - showing four parallel computation streams](/img/tutorials/dask_diagnostic_dashboard.png)

  </TabItem>
  <TabItem value="labextension" label="Dask JupyterLab extension (recommended)">

The [Dask-labextension](https://github.com/dask/dask-labextension) is included with the default Nebari deployment.
It provides a JupyterLab extension to manage Dask clusters,
as well as embed Dask's dashboard plots directly into JupyterLab panes.

On JupyterLab, you can select the Dask icon in the left sidebar and click on the magnifying glass icon to view the list of available plots.
Alternatively, you can copy-and-paste the dashboard link (displayed in the Client UI) in the input field.

![Dask-labextension UI In JupyterLab - displays the demo notebook and a list of cluster settings](/img/tutorials/dask_labextension.png)

  </TabItem>
</Tabs>

Let's open and understand the dashboard plots `Task Stream`, `Progress`, and `Cluster map`.

Most colors and the interpretation would differ based on the computation you choose.
However, the colors remain consistent and represent the same task across all plots within the dashboard.

> Note that some colors always have the same meaning, like red always indicates inter-worker communication and data transfer.

Reviewer: ```suggestion
Note that some colors always have the same meaning. For example, red always indicates inter-worker communication and data transfer.
```

> Each of the computation you submit to Dask (you will learn more in the next section) is split into multiple tasks for parallel execution.

Reviewer: ```suggestion
Each of the computations you submit to Dask (you will learn more in the next section) is split into multiple tasks for parallel execution.
```

> In the `Progress` plot, we see the distinct colors associated with different tasks to complete the overall computation.

Reviewer: ```suggestion
In the `Progress` plot, the distinct colors are associated with different tasks to complete the overall computation.
```

> In the `Task Stream` (a streaming plot) each row represents a thread (a Dask worker can have multiple threads)
> and the small rectangles within are the individual tasks.

Reviewer: ```suggestion
In the `Task Stream` (a streaming plot) each row represents a thread (a Dask worker can have multiple threads). The small rectangles within are the individual tasks running over time.
```

The `Cluster Map` shows the Dask scheduler at the center in a purple circle, with the active Dask workers around it in yellow circles.

> This diagram is a convenient way to visualize Adaptive scaling, and ensure new workers are being spun-up and shut-down based on workflow demand.

Reviewer: ```suggestion
This diagram is a convenient way to visualize Adaptive Scaling, and ensure new workers are being spun-up and shut-down based on workflow demand.
```

Check out the [Dask Documentation on the dashboard plots](https://docs.dask.org/en/stable/dashboard.html) for more information.
Keep the dashboard plots open for the following computations!

## Step 5 - Run your computation with Dask

> In the following sections, you will perform some basic analysis on the well-known New York City yellow taxi dataset,

Reviewer: We need a link to the original source behind "New York City yellow taxi dataset"

Author: Done!

Performing this kind of analysis on such a large dataset would not be possible without a tool like Dask. On Nebari, Dask comes out-of-the-box ready to help you handle these larger-than-memory (out-of-core) datasets.

## Step 5 - Shutdown the cluster

As will you have noticed, you can spin up a lot of compute really quickly using Dask.

**With great power comes great responsibility**

> Remember to shut down your cluster once you are done, otherwise this will be running in the background, and you might incur on unplanned costs.

Reviewer: ```suggestion
Remember to shut down your cluster once you are done, otherwise this will be running in the background, and you might incur unplanned costs.
```

cluster.close(shutdown=True)
```

## Using Dask safely

If you're anything like us, we've forgotten to shut down our cluster a time or two. Wrapping the `dask-gateway` tasks in a

> context manager is a great practice that ensures the cluster is fully shutdown once the task is complete!

Reviewer: I can't comment on that line, but below here on line 311 there is another mention of the dask env name that needs to be updated.

Author: Thanks, done!
